[Gtk4Prep] Viewcontainer cleanup and prepare for Gtk4

jeremypw opened the pull request:
Fixes #2376

Gtk4 Prep:
 - Use event controller
 - No set_events ()

Cleanup:
 - Modernise construction style
 - Modernise layout, reduce line length
 - Merge single use functions

danirabbit commented on src/View/ViewContainer.vala:
```diff
@@ -33,13 +33,13 @@ namespace Files.View {
             }
 
-            set {
+            set construct {
```
Should this be `set construct` or just `construct`? If it's settable I think we need to disconnect signals from the older object before connecting to the new one right?

jeremypw replied:
Dragging a tab to a new or different window causes this property to be set to a different window.  The signals are disconnected in line 38?

jeremypw replied:
On further investigation it turns out there is a pre-existing issue around changing the window associated with a viewcontainer.  This only occurs when dragging and dropping between windows because the viewcontainer is not recreated (it is created new when the new window or open-in actions are used).  I have fixed it here while there is attention on it but it could be done in a separate PR if desired.

danirabbit commented on src/View/ViewContainer.vala:
```diff
@@ -105,34 +138,51 @@ namespace Files.View {
+                button = 0,
+                propagation_phase = CAPTURE
+            };
+            button_controller.pressed.connect ((n_press, x, y) => {
```
Should this be `released` connect instead?

jeremypw replied:
The function it is replacing is a press-event - I do not want to change the UI here.  Please raise an issue if you want a UI change,.

danirabbit commented on src/View/ViewContainer.vala:
```diff
@@ -105,34 +138,51 @@ namespace Files.View {
-                window.folder_deleted.connect (on_folder_deleted);
-            }
+            var button_controller = new Gtk.GestureMultiPress (this) {
+                button = 0,
```
Isn't this `0` by default?

jeremypw replied:
Yes, but I like to make it an explicit choice, at least at this stage.

danirabbit commented on src/View/ViewContainer.vala:
```diff
@@ -105,34 +138,51 @@ namespace Files.View {
-            if (window != null) {
-                window.folder_deleted.connect (on_folder_deleted);
-            }
+            var button_controller = new Gtk.GestureMultiPress (this) {
```
I can't confirm that this works. Is there a reason we add this controller to the viewcontainer widget instead of adding it to the window and triggering `action_back` and `action_forward`? It seems like we'd want these mouse buttons to work when focus is anywhere in the window and not just in the viewcontainer as the keyboard shortcuts do

jeremypw replied:
I didnt think we would want this to work when e.g. the search bar was focused.  It is true that the keyboard shortcut does work in that case but I am not sure that it should. But if you disagree, I can associate this controller with the window instead.  I do not have an appropriate mouse to test this - I'll try to fake a test.

jeremypw replied:
Now moved to window.  I have added additional shortcuts for testing without a special mouse - these can be removed if not wanted.

jeremypw replied:
Ensure you have some paths in both the forward and backwards stacks for testing :smile:

jeremypw commented:
Convert to draft while fixing move tab to new window

jeremypw commented:
@danirabbit Is the special mouse button navigation working in `main`?  That original handler is not getting triggered for me on button press any more.  However, the button controller attached to the window *does* work for me.  However, I will revert that change and work on the issue in a separate PR.

jeremypw commented:
@danirabbit I have reverted changes to button press handling. I have also fixed the terminal warnings caused by associating the viewcontainer and the window later in order to fix #2376.  Significant simplification/clarification of ViewContainer was achieved out in the Gtk4 port `files4`branch  but leaving that for later.  Not sure whether should be done before or after the port (or at all).

danirabbit commented:
@jeremypw mouse buttons are working in main for me and I can confirm they are now working in this branch after reverting.

I noticed another new terminal message when secondary clicking in the view:
```
** (io.elementary.files:68233): CRITICAL **: 08:47:07.258: files_view_window_can_bookmark_uri: assertion 'self != NULL' failed
```

jeremypw commented:
@danirabbit That's strange because buttons no longer work for me (except when handled by the view rather than the view container (using touchpad buttons). The viewcontainer handler is never triggered.  I'll try and find out what is going on in another PR.  Meanwhile I'll fix that terminal message.

jeremypw commented:
@danirabbit Fixed terminal warning and checked code recognizing that the viewcontainer window property is now 'null' on construction and only becomes available when it is associated with a tab-view.  Where there is no expectation that the window would be null at the time a function is called a `requires (window != null)` is added to make this explicit.  Where necessary a check is performed.

jeremypw commented:
I probably should have left fixing that issue to a later PR :frowning_face:  - it has caused a lot of regressions cause of the timing change.

jeremypw commented:
@danirabbit Now the directory loads after the window property is set (should have done that in the first place!) . This should fix all the regressions.

danirabbit commented:
I've got a reproducible crash here. Steps are kind of complicated:

1. right click and select "Open in new Window"
2. Close the window
3. right click the tab and select "duplicate tab"
4. Drag the duplicated tab out of the window

I can't reproduce in main

backtrace
```
Starting program: /home/dani/Projects/files/build/src/io.elementary.files 
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff2fff640 (LWP 13653)]
[New Thread 0x7ffff27fe640 (LWP 13654)]
[New Thread 0x7ffff1feb640 (LWP 13655)]
[New Thread 0x7ffff16ae640 (LWP 13656)]
[New Thread 0x7fffe3fff640 (LWP 13658)]
[Thread 0x7fffe3fff640 (LWP 13658) exited]
[New Thread 0x7fffe3fff640 (LWP 13659)]
[New Thread 0x7fffe37fe640 (LWP 13660)]
[Thread 0x7fffe3fff640 (LWP 13659) exited]
[Thread 0x7fffe37fe640 (LWP 13660) exited]
[New Thread 0x7fffe37fe640 (LWP 13661)]
[New Thread 0x7fffe3fff640 (LWP 13662)]
[Thread 0x7fffe37fe640 (LWP 13661) exited]
[Thread 0x7fffe3fff640 (LWP 13662) exited]
[New Thread 0x7fffe3fff640 (LWP 13663)]
[New Thread 0x7fffe37fe640 (LWP 13664)]
[Thread 0x7fffe3fff640 (LWP 13663) exited]
[Thread 0x7fffe37fe640 (LWP 13664) exited]
[New Thread 0x7fffe37fe640 (LWP 13665)]
[New Thread 0x7fffe3fff640 (LWP 13666)]
[New Thread 0x7fffd7bff640 (LWP 13668)]
[New Thread 0x7fffd73fe640 (LWP 13669)]
[New Thread 0x7fffd6bfd640 (LWP 13671)]
[New Thread 0x7fffd63fc640 (LWP 13672)]
[Thread 0x7fffd6bfd640 (LWP 13671) exited]
[Thread 0x7fffd63fc640 (LWP 13672) exited]
[Thread 0x7fffd73fe640 (LWP 13669) exited]
[Thread 0x7ffff16ae640 (LWP 13656) exited]
[Thread 0x7fffe37fe640 (LWP 13665) exited]
[Thread 0x7fffe3fff640 (LWP 13666) exited]
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug

*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug


(io.elementary.files:13650): Gtk-CRITICAL **: 10:52:55.669: gtk_widget_queue_draw_area: assertion 'width >= 0' failed
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug


(io.elementary.files:13650): Gtk-CRITICAL **: 10:52:55.669: gtk_widget_queue_draw_area: assertion 'width >= 0' failed

(io.elementary.files:13650): Gtk-CRITICAL **: 10:52:55.669: gtk_widget_queue_draw_area: assertion 'width >= 0' failed
*** BUG ***
In pixman_region32_init_rect: Invalid rectangle passed
Set a breakpoint on '_pixman_log_error' to debug


(io.elementary.files:13650): Gtk-CRITICAL **: 10:52:55.682: gtk_widget_queue_draw_area: assertion 'width >= 0' failed

(io.elementary.files:13650): Gtk-CRITICAL **: 10:52:55.683: gtk_widget_queue_draw_area: assertion 'width >= 0' failed
[New Thread 0x7fffe3fff640 (LWP 13673)]
[New Thread 0x7fffe37fe640 (LWP 13674)]
[New Thread 0x7ffff16ae640 (LWP 13675)]
[New Thread 0x7fffd73fe640 (LWP 13676)]
[New Thread 0x7fffd6bfd640 (LWP 13677)]
[New Thread 0x7fffd63fc640 (LWP 13678)]
[New Thread 0x7fffd5bfb640 (LWP 13679)]
[New Thread 0x7fffd53fa640 (LWP 13680)]
[New Thread 0x7fffd4bf9640 (LWP 13681)]
[New Thread 0x7fffb3fff640 (LWP 13682)]
[New Thread 0x7fffab7fe640 (LWP 13683)]
[New Thread 0x7fffb37fe640 (LWP 13684)]
[New Thread 0x7fffb2ffd640 (LWP 13685)]
[New Thread 0x7fffb27fc640 (LWP 13686)]
[New Thread 0x7fffb1ffb640 (LWP 13687)]
[New Thread 0x7fffb17fa640 (LWP 13688)]
[New Thread 0x7fffb0ff9640 (LWP 13689)]
[Thread 0x7fffb0ff9640 (LWP 13689) exited]
[Thread 0x7fffb17fa640 (LWP 13688) exited]
[Thread 0x7fffb1ffb640 (LWP 13687) exited]
[Thread 0x7fffb2ffd640 (LWP 13685) exited]
[Thread 0x7fffb37fe640 (LWP 13684) exited]
[Thread 0x7fffab7fe640 (LWP 13683) exited]
[Thread 0x7fffb3fff640 (LWP 13682) exited]
[New Thread 0x7fffb3fff640 (LWP 13690)]
[New Thread 0x7fffab7fe640 (LWP 13691)]
[New Thread 0x7fffb0ff9640 (LWP 13692)]
[New Thread 0x7fffb2ffd640 (LWP 13693)]
[New Thread 0x7fffb37fe640 (LWP 13694)]
[New Thread 0x7fffb1ffb640 (LWP 13695)]
[Thread 0x7fffb1ffb640 (LWP 13695) exited]
[Thread 0x7fffb37fe640 (LWP 13694) exited]
[Thread 0x7fffb2ffd640 (LWP 13693) exited]
[Thread 0x7fffb0ff9640 (LWP 13692) exited]
[Thread 0x7fffab7fe640 (LWP 13691) exited]
[Thread 0x7fffb3fff640 (LWP 13690) exited]
[Thread 0x7fffb27fc640 (LWP 13686) exited]
[Thread 0x7fffe3fff640 (LWP 13673) exited]
[New Thread 0x7fffb27fc640 (LWP 13696)]
[Thread 0x7fffb27fc640 (LWP 13696) exited]
[New Thread 0x7fffb27fc640 (LWP 13697)]
[New Thread 0x7fffe3fff640 (LWP 13698)]
[Thread 0x7fffb27fc640 (LWP 13697) exited]
[Thread 0x7fffe3fff640 (LWP 13698) exited]

(io.elementary.files:13650): Handy-CRITICAL **: 10:52:59.513: hdy_tab_view_get_page_position: assertion 'HDY_IS_TAB_PAGE (page)' failed

** (io.elementary.files:13650): CRITICAL **: 10:52:59.521: files_view_view_container_get_uri: assertion 'self != NULL' failed

** (io.elementary.files:13650): CRITICAL **: 10:52:59.521: files_view_window_update_labels: assertion 'uri != NULL' failed

** (io.elementary.files:13650): CRITICAL **: 10:52:59.521: files_view_view_container_set_active_state: assertion 'self != NULL' failed

** (io.elementary.files:13650): CRITICAL **: 10:52:59.521: files_view_view_container_get_uri: assertion 'self != NULL' failed

** (io.elementary.files:13650): CRITICAL **: 10:52:59.522: sidebar_sidebar_window_real_sync_uri: assertion 'location != NULL' failed

** (io.elementary.files:13650): CRITICAL **: 10:52:59.522: files_view_view_container_get_is_frozen: assertion 'self != NULL' failed

(io.elementary.files:13650): Handy-CRITICAL **: 10:52:59.522: hdy_tab_view_get_page_position: assertion 'HDY_IS_TAB_PAGE (page)' failed

(io.elementary.files:13650): Handy-CRITICAL **: 10:52:59.528: hdy_tab_view_get_page_position: assertion 'HDY_IS_TAB_PAGE (page)' failed
[New Thread 0x7fffe3fff640 (LWP 13699)]
[New Thread 0x7fffb27fc640 (LWP 13700)]

(io.elementary.files:13650): Gtk-CRITICAL **: 10:52:59.701: gtk_widget_queue_draw_area: assertion 'width >= 0' failed

(io.elementary.files:13650): Gtk-CRITICAL **: 10:52:59.750: gtk_widget_queue_draw_area: assertion 'width >= 0' failed

(io.elementary.files:13650): Gtk-CRITICAL **: 10:52:59.751: gtk_widget_queue_draw_area: assertion 'width >= 0' failed
[Thread 0x7fffe3fff640 (LWP 13699) exited]
[Thread 0x7fffb27fc640 (LWP 13700) exited]
[New Thread 0x7fffb27fc640 (LWP 13701)]
[Thread 0x7fffd7bff640 (LWP 13668) exited]

(io.elementary.files:13650): Gtk-CRITICAL **: 10:53:03.522: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(io.elementary.files:13650): Gtk-WARNING **: 10:53:03.522: gtk_widget_size_allocate(): attempt to allocate widget with width 30 and height -6

(io.elementary.files:13650): Gtk-WARNING **: 10:53:03.522: Negative content height -7 (allocation 1, extents 4x4) while allocating gadget (node button, owner GtkButton)

(io.elementary.files:13650): Gtk-CRITICAL **: 10:53:03.522: gtk_widget_get_preferred_width_for_height: assertion 'height >= 0' failed

(io.elementary.files:13650): Gtk-WARNING **: 10:53:03.522: gtk_widget_size_allocate(): attempt to allocate widget with width 34 and height -6

(io.elementary.files:13650): Gtk-CRITICAL **: 10:53:03.669: gtk_widget_queue_draw_area: assertion 'width >= 0' failed

(io.elementary.files:13650): Gtk-CRITICAL **: 10:53:03.686: gtk_widget_queue_draw_area: assertion 'width >= 0' failed

(io.elementary.files:13650): Gtk-CRITICAL **: 10:53:03.686: gtk_widget_queue_draw_area: assertion 'width >= 0' failed

(io.elementary.files:13650): GLib-GObject-WARNING **: 10:53:03.763: ../../../gobject/gsignal.c:2765: instance '0x5555561655b0' has no handler with id '6218'

(io.elementary.files:13650): GLib-GObject-CRITICAL **: 10:53:03.763: g_object_ref: assertion '!object_already_finalized' failed

Thread 1 "io.elementary.f" received signal SIGSEGV, Segmentation fault.
0x00007ffff7d270de in ?? () from /lib/x86_64-linux-gnu/libgobject-2.0.so.0
(gdb) bt
#0  0x00007ffff7d270de in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#1  0x00007ffff7d24d2f in g_closure_invoke () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#2  0x00007ffff7d40ba0 in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#3  0x00007ffff7d42614 in g_signal_emit_valist () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#4  0x00007ffff7d42863 in g_signal_emit () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#5  0x00007ffff75a7600 in  () at /lib/x86_64-linux-gnu/libgtk-3.so.0
#6  0x00007ffff75b70ee in  () at /lib/x86_64-linux-gnu/libgtk-3.so.0
#7  0x00007ffff732126f in  () at /lib/x86_64-linux-gnu/libgtk-3.so.0
#8  0x00007ffff7d32ed1 in g_object_unref () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#9  0x0000555555605b13 in block47_data_unref (_userdata_=0x555555864a80) at ../src/View/Window.vala:188
#10 0x00007ffff7d245f8 in g_closure_unref () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#11 0x00007ffff7d40098 in g_signal_handlers_destroy () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#12 0x00007ffff7d2f3a1 in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#13 0x00007ffff7d32ed1 in g_object_unref () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#14 0x00007ffff7f7a557 in files_view_chrome_view_switcher_finalize (obj=0x555556028680) at ../libcore/Widgets/ViewSwitcher.vala:26
#15 0x00007ffff7d32f9d in g_object_unref () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#16 0x0000555555615f92 in files_view_window_finalize (obj=0x555555a1e830) at ../src/View/Window.vala:84
#17 0x00007ffff7d32f9d in g_object_unref () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#18 0x00007ffff7f3fc10 in files_plugins_base_interface_loaded (self=0x555555939e50, widget=0x555555a1ebc0) at ../libcore/Plugin.vala:29
#19 0x00007ffff7f3eb6a in files_plugin_manager_interface_loaded (self=0x55555594d210, win=0x555555a1ebc0) at ../libcore/PluginManager.vala:221
#20 0x000055555558c1ae in files_application_real_window_added (base=0x5555556c8130, window=0x555555a1ebc0) at ../src/Application.vala:227
#21 0x00007ffff7d24d2f in g_closure_invoke () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#22 0x00007ffff7d40895 in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#23 0x00007ffff7d42614 in g_signal_emit_valist () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#24 0x00007ffff7d42863 in g_signal_emit () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#25 0x00007ffff75b69e1 in gtk_window_set_application () at /lib/x86_64-linux-gnu/libgtk-3.so.0
#26 0x00007ffff7d34d86 in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#27 0x00007ffff7d350b0 in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#28 0x00007ffff7d368f8 in g_object_new_valist () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#29 0x00007ffff7d36e2d in g_object_new () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#30 0x0000555555605a46 in files_view_window_constructPython Exception <class 'gdb.error'>: No type named TypeNode.
 (object_type=, _application=0x5555556c8130) at ../src/View/Window.vala:104
#31 0x0000555555605a7b in files_view_window_new (_application=0x5555556c8130) at ../src/View/Window.vala:103
#32 0x000055555560657b in __lambda201_ (self=0x555555a1e4a0) at ../src/View/Window.vala:375
#33 0x00005555556065f0 in ___lambda201__hdy_tab_view_create_window (_sender=0x555555cd2140, self=0x555555a1e4a0) at /home/dani/Projects/files/build/Window.c:1763
#34 0x00007ffff6c38e2e in  () at /lib/x86_64-linux-gnu/libffi.so.8
#35 0x00007ffff6c35493 in  () at /lib/x86_64-linux-gnu/libffi.so.8
#36 0x00007ffff7d2b69a in g_cclosure_marshal_generic_va () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#37 0x00007ffff7d42700 in g_signal_emit_valist () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#38 0x00007ffff7d42863 in g_signal_emit () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#39 0x00007ffff6e91dd3 in  () at /lib/x86_64-linux-gnu/libhandy-1.so.0
#40 0x00007ffff6e91f78 in  () at /lib/x86_64-linux-gnu/libhandy-1.so.0
#41 0x00007ffff75e64bb in  () at /lib/x86_64-linux-gnu/libgtk-3.so.0
#42 0x00007ffff7d24d2f in g_closure_invoke () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#43 0x00007ffff7d40624 in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
--Type <RET> for more, q to quit, c to continue without paging--
#44 0x00007ffff7d42026 in g_signal_emit_valist () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#45 0x00007ffff7d42a8e in g_signal_emit_by_name () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#46 0x00007ffff75cb736 in  () at /lib/x86_64-linux-gnu/libgtk-3.so.0
#47 0x00007ffff7d24d2f in g_closure_invoke () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#48 0x00007ffff7d40c36 in  () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#49 0x00007ffff7d42614 in g_signal_emit_valist () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#50 0x00007ffff7d42863 in g_signal_emit () at /lib/x86_64-linux-gnu/libgobject-2.0.so.0
#51 0x00007ffff7169750 in  () at /lib/x86_64-linux-gnu/libgdk-3.so.0
#52 0x00007ffff713271e in  () at /lib/x86_64-linux-gnu/libgdk-3.so.0
#53 0x00007ffff7169f56 in  () at /lib/x86_64-linux-gnu/libgdk-3.so.0
#54 0x00007ffff7dc5d3b in g_main_context_dispatch () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#55 0x00007ffff7e1b258 in  () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#56 0x00007ffff7dc33e3 in g_main_context_iteration () at /lib/x86_64-linux-gnu/libglib-2.0.so.0
#57 0x00007ffff7c17fb5 in g_application_run () at /lib/x86_64-linux-gnu/libgio-2.0.so.0
#58 0x0000555555597721 in _vala_main (args=0x7fffffffdd88, args_length1=1) at ../src/main.vala:28
#59 0x0000555555597768 in main (argc=1, argv=0x7fffffffdd88) at ../src/main.vala:16
```

jeremypw commented:
Weird!  You can drag a duplicate tab out of the window OK without the prior steps. Investigating.

jeremypw commented:
@danirabbit Turned out to be an obscure bug related to use of the experimental `with () {}` syntax.  When that is replaced the bug does not occur.  No idea why it does not occur in master though.  Something to do with a race between refs being added and removed.

danirabbit reviewed (approved):
Definitely fixed :rocket:
